File: rust/src/bin/p0021.rs
// Project Euler: Problem 21

euler::run_solver!(21);

fn solve() -> String {
    compute(10_000).to_string()
}

fn compute(limit: usize) -> u64 {
    use euler::math;

    debug_assert!(limit > 1);

    let d_tbl = math::aliquot_sum_tbl(limit - 1);

    (1..limit)
        .filter(|x| *x as u64 > d_tbl[*x] && d_tbl[d_tbl[*x] as usize] == *x as u64)
        .map(|x| x as u64 + d_tbl[x])
        .sum()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0021_10000() {
        assert_eq!(compute(10_000), 31626);
    }
}

File: ocaml/src/p0021.ml
(* Project Euler: Problem 21 *)

open Core

let compute limit =
  let spd_tbl = Euler.Math.aliquot_sum_tbl limit in
  List.(
    range 2 limit ~stop:`exclusive
    |> filter_map ~f:(fun x ->
      let y = spd_tbl.(x) in
      if x > y && spd_tbl.(y) = x then Some (x + y) else None)
    |> reduce_exn ~f:( + ))
;;

let solve () = compute 10_000 |> Int.to_string

(* Test *)

let%test_unit "10_000" = [%test_eq: int] (compute 10_000) 31626
